builder/dockerfile/builder.go

```diff
@@ -168,62 +177,78 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
 	}
 	return aux.Emit(types.BuildResult{ID: state.imageID})
 }
+func convertMapToEnvs(m map[string]string) []string {
+	result := []string{}
+	for k, v := range m {
+		result = append(result, k+"="+v)
+	}
+	return result
+}
+func (b *Builder) processMetaArg(meta instructions.ArgCommand, shlex *ShellLex) error {
+	envs := convertMapToEnvs(b.buildArgs.GetAllAllowed())
+	if err := meta.Expand(func(word string) (string, error) {
+		return shlex.ProcessWord(word, envs)
+	}); err != nil {
+		return err
+	}
+	b.buildArgs.AddArg(meta.Name, meta.Value)
+	b.buildArgs.AddMetaArg(meta.Name, meta.Value)
```

Comment:
Hmm I like the idea (and it is required for build parallelisation). We still need a common buildArgs instance for metaArgs construction, but once it is done, we should simply clone it once per stage, and have individual stage have their own instance.